File: frontend/src/types/task.ts
/** Kanban column status for a task — matches backend TaskStatus enum. */
export type TaskStatus = 'backlog' | 'todo' | 'in_progress' | 'review' | 'done';

/** All valid TaskStatus values in Kanban flow order. */
export const TASK_STATUSES: TaskStatus[] = ['backlog', 'todo', 'in_progress', 'review', 'done'];

/** Valid Kanban transitions per column. */
export const VALID_TASK_TRANSITIONS: Record<TaskStatus, TaskStatus[]> = {
  backlog: ['backlog', 'todo'],
  todo: ['backlog', 'in_progress'],
  in_progress: ['todo', 'review'],
  review: ['in_progress', 'done'],
  done: ['done'],
};

/** Check if a transition from current to next status is valid. */
export function isValidTaskTransition(current: TaskStatus, next: TaskStatus): boolean {
  return VALID_TASK_TRANSITIONS[current]?.includes(next) ?? false;
}

/** Get allowed transitions from a given status. */
export function getAllowedTaskTransitions(current: TaskStatus): TaskStatus[] {
  return VALID_TASK_TRANSITIONS[current] ?? [];
}

export interface Task {
  id: string;
  storyId: string;
  title: string;
  description: string;
  labels: string[];
  dependencies: string[];
  status: TaskStatus;
  priority: string;
  createdAt: string;
  updatedAt: string;
}

/** Raw task from API (snake_case). */
export interface RawTaskItem {
  id: string;
  user_story_id: string;
  title: string;
  description: string;
  status: TaskStatus;
  priority: string;
  labels: string[];
  dependencies: string[];
  created_at: string;
  updated_at: string;
}

File: backend/src/storico/domain/entities/task.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from enum import StrEnum
from uuid import UUID

from uuid_utils.compat import uuid7


class TaskStatus(StrEnum):
    """Kanban column status for a task."""

    BACKLOG = "backlog"
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    DONE = "done"


@dataclass(frozen=True, slots=True)
class Task:
    user_story_id: UUID
    title: str
    description: str = ""
    status: TaskStatus = field(default=TaskStatus.BACKLOG)
    priority: str = "medium"
    labels: list[str] = field(default_factory=list)
    dependencies: list[str] = field(default_factory=list)
    id: UUID = field(default_factory=uuid7)
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))
